### DISCUSSION TITLE
sqlalchemy.engine.URL doesn't support dialects, so what's the recommended way to add them?

### DISCUSSION THREAD

**Discussion by dougthor42** [sqlalchemy/sqlalchemy#6475] [Usage Questions] [answered]
`sqlalchemy.engine.URL` (`sqlalchemy.engine.url.URL` before v1.4) doesn't support dialects? I'm kinda surprised, given that [RFC-1738](https://datatracker.ietf.org/doc/html/rfc1738) allows "+" in the `scheme` part of the URL (see section 5).

If we're not *supposed* to use `sqlalchemy.engine.URL` for dialects, what's the recommended way of setting the dialect?

What I've been doing is either:

1. Creating the URL object, wrapping it in `str`, and then prepending my dialect, or
2. Making my own MyURL object which inherits from engine.URL and overriding `__str__` and `__repr__` so that they include the dialect.

### Example of what I would expect to work:

+ Python 3.6
+ SQLAlchemy 1.3.24
+ PyMySQL 0.8.1

```python
# foo.py
from sqlalchemy.engine.url import URL

url = URL(
    drivername="pymysql",
    username="user",
    password="password",
    host="host",
    port=3306,
    database="bar",
    dialect="mysql",
)
print(url)
```

#### Got:

```
$ python foo.py
Traceback (most recent call last):
  File "foo.py", line 10, in <module>
    dialect="mysql",
TypeError: __init__() got an unexpected keyword argument 'dialect'
```

#### Expected:

```
mysql+pymysql://user:password@host:3306/bar
```

**Comment by CaselIT** [comment, marked as answer]
Hi,

Pleas refer to the [documentation](https://docs.sqlalchemy.org/en/14/core/engines.html?highlight=url#sqlalchemy.engine.URL) for usage info.
Regarding the RFC, I don't think URL is expect to follow that standard, or any other for that matter.

Your example would be
```py
from sqlalchemy.engine.url import URL

url = URL.create(
    drivername="mysql+pymysql",
    username="user",
    password="password",
    host="host",
    port=3306,
    database="bar",
)
print(url)
```

**Reply by zzzeek** [reply]
noticed you just updated that "var", im also changing those to attr because they seem to not be rendering now, not sure if this is sphinx 1.4 or it was always broken

**Reply by CaselIT** [reply]
I also noticed it after pushing that change

**Comment by zzzeek** [comment]
"drivername" is the dialect + driver together, the [docs](https://docs.sqlalchemy.org/en/14/core/engines.html?highlight=url%20url#sqlalchemy.engine.URL) in 1.4 should be pretty clear on this but there is no difference vs. 1.3 

```python
>>> import sqlalchemy
sq>>> sqlalchemy.__version__
'1.3.25'
>>> from sqlalchemy.engine.url import URL
>>> url = URL(
...     drivername="mysql+pymysql",
...     username="user",
...     password="password",
...     host="host",
...     port=3306,
...     database="bar",
... )
>>> print(url)
mysql+pymysql://user:password@host:3306/bar
```

**Reply by CaselIT** [reply]
ah we both replied the same. btw there was a typo in the URL docs 66cb4e99

**Comment by dougthor42** [comment]
Ah, you're both right! I was looking at the docs for **1.3**, which doesn't mention it:
![image](https://user-images.githubusercontent.com/5386897/118049920-4d81fa80-b333-11eb-880e-d1384537299a.png)


But the docs for 1.4 does:
![image](https://user-images.githubusercontent.com/5386897/118049958-5e327080-b333-11eb-9dd6-2dc7212d135b.png)
